crates/ursa-network/src/config.rs

```diff
@@ -2,65 +2,102 @@ use libp2p::Multiaddr;
 use serde::{Deserialize, Serialize};
 use std::path::PathBuf;
 
-pub const DEFAULT_TRACKER_URL: &str = "https://tracker.ursa.earth/register";
-pub const DEFAULT_BOOTSTRAP: [&str; 2] = [
-    "/ip4/159.223.211.234/tcp/6009/p2p/12D3KooWDji7xMLia6GAsyr4oiEFD2dd3zSryqNhfxU3Grzs1r9p",
-    "/ip4/146.190.232.131/tcp/6009/p2p/12D3KooWGw8vCj9XayJDMXUiox6pCUFm7oVuWkDJeE2H9SDQVEcM",
-];
-pub const DEFAULT_DB_PATH_STR: &str = "~/.ursa/data/ursa_db";
-pub const DEFAULT_KEYSTORE_PATH_STR: &str = "~/.ursa/keystore";
-
 /// Ursa Configuration
 #[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
 pub struct NetworkConfig {
     /// Optional mdns local discovery.
+    #[serde(default = "NetworkConfig::default_mdns")]
     pub mdns: bool,
     /// Optional Provide a relay server for other peers to listen on.
+    #[serde(default = "NetworkConfig::default_relay_server")]
     pub relay_server: bool,
     /// Optional autonat. This is used to determine if we are behind a NAT and need to use a relay.
+    #[serde(default = "NetworkConfig::default_autonat")]
     pub autonat: bool,
     /// Optional Enable listening on a relay server if not publicly available. Requires autonat.
     /// Connections will attempt to upgrade using dcutr.
+    #[serde(default = "NetworkConfig::default_relay_client")]
     pub relay_client: bool,
     /// set true if it is a bootstrap node. default = false
+    #[serde(default = "NetworkConfig::default_bootstrapper")]
     pub bootstrapper: bool,
     /// Swarm listening Address.
+    #[serde(default = "NetworkConfig::default_swarm_addrs")]
     pub swarm_addrs: Vec<Multiaddr>,
     /// Bootstrap nodes.
+    #[serde(default = "NetworkConfig::default_bootstrap_nodes")]
     pub bootstrap_nodes: Vec<Multiaddr>,
     /// Database path.
+    #[serde(default = "NetworkConfig::default_database_path")]
     pub database_path: PathBuf,
     /// user identity name
+    #[serde(default = "NetworkConfig::default_identity")]
     pub identity: String,
     /// Keystore path. Defaults to ~/.ursa/keystore
+    #[serde(default = "NetworkConfig::default_keystore_path")]
     pub keystore_path: PathBuf,
     /// Temporary HTTP tracker url. This is used for pre-consensus node registrations.
     /// Defaults to devnet tracker.
+    #[serde(default = "NetworkConfig::default_tracker")]
     pub tracker: String,
 }
 
+impl NetworkConfig {
+    fn default_mdns() -> bool {
```

Comment:
Can we not implement a default trait much more readable here instead of a function returning a truth value

Reply:
Default is implemented just under it, but we can't use individual fields from there in the derive macro

Reply:
So a complexity was introduced? We can use defaults from a struct then tried to alter it by introducing other functions. I am not sure I follow the logic here. Additionally a value can used and as I mentioned a default trait can be created check here  https://serde.rs/attr-default.html

Reply:
have a read of that document. You can only provide custom defaults through a function as mentioned in my other response

Reply:
Have a read at our old code where we use defaults and we should stick to that

Reply:
If i'm thinking of what you're saying correctly, he old code did use Option<T> on some fields and insert defaults after the fact, but that isn't on deserialization but rather usage and we'd need to make all current and future fields in the config struct Option<T> and fill them with Some(T) when loading the config, which requires unwraps or filling into a secondary struct to use throughout the codebase nicely. Whereas we can just call a defined function to fill the default value and directly use the fields without unwraps or additional handling